tfe/resource_tfe_team_members.go

```diff
@@ -0,0 +1,167 @@
+package tfe
+
+import (
+	"fmt"
+	"log"
+
+	tfe "github.com/hashicorp/go-tfe"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceTFETeamMembers() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceTFETeamMembersCreate,
+		Read:   resourceTFETeamMembersRead,
+		Update: resourceTFETeamMembersUpdate,
+		Delete: resourceTFETeamMembersDelete,
+
+		Schema: map[string]*schema.Schema{
+			"team_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"usernames": &schema.Schema{
+				Type:     schema.TypeSet,
+				Required: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+		},
+	}
+}
+
+func resourceTFETeamMembersCreate(d *schema.ResourceData, meta interface{}) error {
+	tfeClient := meta.(*tfe.Client)
+
+	// Get the team ID and username..
+	teamID := d.Get("team_id").(string)
+
+	// Create a new options struct.
+	options := tfe.TeamMemberAddOptions{}
+
+	// Add all the users that need to be added.
+	for _, username := range d.Get("usernames").(*schema.Set).List() {
+		options.Usernames = append(options.Usernames, username.(string))
+	}
+
+	log.Printf("[DEBUG] Add users to team: %s", teamID)
+	err := tfeClient.TeamMembers.Add(ctx, teamID, options)
+	if err != nil {
+		return fmt.Errorf("Error adding users to team %s: %v", teamID, err)
+	}
+
+	d.SetId(teamID)
+
+	return nil
+}
+
+func resourceTFETeamMembersRead(d *schema.ResourceData, meta interface{}) error {
+	tfeClient := meta.(*tfe.Client)
+
+	// Get the team ID and username..
+	teamID := d.Get("team_id").(string)
+
+	log.Printf("[DEBUG] Read users from team: %s", teamID)
+	users, err := tfeClient.TeamMembers.List(ctx, teamID)
+	if err != nil {
+		if err == tfe.ErrResourceNotFound {
+			log.Printf("[DEBUG] Users do no longer exist")
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error reading users from team %s: %v", teamID, err)
+	}
+
+	usernames := resourceTFETeamMembers().Schema["usernames"].ZeroValue().(*schema.Set)
```

Comment:
why can't we just use a normal field/Set here? e.g.

```
usernames := make([]string, 0)
for _, user := range users {
  usernames = append(user.Username)
}

if err := d.Set("usernames", usernames); err != nil {
  return fmt.Errorf("Error setting `usernames`: %+v", err)
}
```

Reply:
Nice one! I developed this pattern back when `d.setWriter.WriteField` (and all related code) didn't exist yet and (IIRC) it was needed then to make sure the hashes were generated correctly.

But looking at that code now, this is indeed totally unnecessary 👍 Fixed...